File: core/src/main/java/xtdb/vector/extensions/UriVector.java
package xtdb.vector.extensions;

import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.vector.VarCharVector;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.FieldType;

import java.net.URI;
import java.nio.charset.StandardCharsets;

public class UriVector extends XtExtensionVector<VarCharVector> {
    public UriVector(String name, BufferAllocator allocator, FieldType fieldType) {
        super(name, allocator, fieldType, new VarCharVector(name, allocator));
    }

    public UriVector(Field field, BufferAllocator allocator) {
        super(field, allocator, new VarCharVector(field, allocator));
    }

    @Override
    public URI getObject(int index) {
        String s = new String(getUnderlyingVector().get(index), StandardCharsets.UTF_8);
        return URI.create(s);
    }
}

File: core/src/main/kotlin/xtdb/vector/extensions/UriVector.kt
package xtdb.vector.extensions

import org.apache.arrow.memory.BufferAllocator
import org.apache.arrow.vector.VarCharVector
import org.apache.arrow.vector.types.pojo.FieldType
import java.net.URI
import java.nio.charset.StandardCharsets

class UriVector(name: String, allocator: BufferAllocator, fieldType: FieldType) :
    XtExtensionVector<VarCharVector>(name, allocator, fieldType, VarCharVector(name, allocator)) {

    init {
        require(fieldType.type == UriType)
    }

    override fun getObject0(index: Int): URI = URI.create(String(underlyingVector[index], StandardCharsets.UTF_8))
}
